pyfar/utils.py

```diff
@@ -0,0 +1,129 @@
+import pyfar as pf
+import numpy as np
+
+
+def broadcast_cshape(signal, cshape):
+    """
+    Broadcast signal to cshape.
+
+    The :py:mod:`cshape <pyfar._concepts.audio_classes>` of the signal is
+    broadcasted following the `numpy broadcasting rules
+    <https://numpy.org/doc/stable/user/basics.broadcasting.html>`_
+
+    Parameters
+    ----------
+    signal : Signal, TimeData, FrequencyData
+        The signal to be broadcasted.
+    cshape : tuple
+        The cshape to which `signal` is broadcasted.
+
+    Returns
+    -------
+    signal : Signal, TimeData, FrequencyData
+        The broadcasted input signal
+    """
+
+    if not isinstance(signal, (pf.Signal, pf.TimeData, pf.FrequencyData)):
+        raise TypeError("All input data must be pyfar audio objects")
+
+    signal = signal.copy()
+    signal._data = np.broadcast_to(
+        signal._data, cshape + (signal._data.shape[-1], ))
+    return signal
+
+
+def broadcast_cshapes(signals, cshape=None):
+    """
+    Broadcast multiple signals to a common cshape.
+
+    The :py:mod:`cshape <pyfar._concepts.audio_classes>` of the signals are
+    broadcasted following the `numpy broadcasting rules
+    <https://numpy.org/doc/stable/user/basics.broadcasting.html>`_
+
+    Parameters
+    ----------
+    signals : tuple of Signal, TimeData, and FrequencyData objects
+        The signals to be broadcasted.
+    cshape : tuple, optional
+        The cshape to which the signals are broadcasted. If `chsape` is
+        ``None`` it is determined from the cshapes of the input signals using
+        ``numpy.broadcast_shapes``. The default is ``None``.
+
+    Returns
+    -------
+    signals : tuple of Signal, TimeData, and FrequencyData objects
+        The broadcasted input signals
+    """
+
+    for signal in signals:
+        if not isinstance(signal, (pf.Signal, pf.TimeData, pf.FrequencyData)):
+            raise TypeError("All input data must be pyfar audio objects")
+
+    if cshape is None:
+        cshape = np.broadcast_shapes(*[s.cshape for s in signals])
+    return [broadcast_cshape(s, cshape) for s in signals]
+
+
+def broadcast_cdim(signal, cdim):
+    """
+    Broadcast signal to cdim.
+
+    The channel dimension (cdim) is the length of the
+    :py:mod:`cshape <pyfar._concepts.audio_classes>` of the signal. The signal
+    is broadcasted to `cdim` by prepending ``cdim - len(signal.cshape)``
```

Comment:
Could it be useful to append instead of prepending?

Reply:
This is following the standard numpy rules for broadcasting. Thats why I did not add an option for this.

Reply:
ok, thanks for clarifying